sdks/python/apache_beam/transforms/periodicsequence.py

```diff
@@ -173,31 +216,92 @@ class PeriodicImpulse(PTransform):
+        self.stop_ts = Timestamp.of(end)
+      else:
+        end = self.stop_ts.micros / 1000000
+    else:
+      end = self.stop_ts
+
+    # The total time for the impulse signal which occurs in [start, end).
+    impulse_duration = end - start
+    if data_duration + self.interval < impulse_duration:
+      # We don't have enough data for the impulse.
+      # If we can fit at least one more data point in the impulse duration,
+      # then we will be in the repeat mode.
+      message = 'The number of elements in the provided pre-timestamped ' \
+        'data sequence is not enough to span the full impulse duration. ' \
+        f'Expected duration: {impulse_duration:.6f}, ' \
+        f'actual data duration: {data_duration:.6f}.'
+
+      if is_pre_timestamped:
+        raise ValueError(
+            f'{message} Please either provide more data or decrease '
+            '`stop_timestamp`.')
+      else:
+        warnings.warn(
+            f'{message} As a result, the data sequence will be repeated to '
+            'generate elements for the entire duration.')
+
   def __init__(
       self,
-      start_timestamp=Timestamp.now(),
-      stop_timestamp=MAX_TIMESTAMP,
-      fire_interval=360.0,
-      apply_windowing=False):
+      start_timestamp: Union[Timestamp, float] = Timestamp.now(),
+      stop_timestamp: Union[Timestamp, float] = MAX_TIMESTAMP,
+      fire_interval: float = 360.0,
+      apply_windowing: bool = False,
+      data: Optional[Sequence[Any]] = None):
     '''
     :param start_timestamp: Timestamp for first element.
     :param stop_timestamp: Timestamp after which no elements will be output.
     :param fire_interval: Interval in seconds at which to output elements.
     :param apply_windowing: Whether each element should be assigned to
       individual window. If false, all elements will reside in global window.
+    :param data: The sequence of elements to emit into the PCollection.
+      The elements can be raw values or pre-timestamped tuples in the format
+      `(apache_beam.utils.timestamp.Timestamp, value)`.
     '''
     self.start_ts = start_timestamp
     self.stop_ts = stop_timestamp
     self.interval = fire_interval
     self.apply_windowing = apply_windowing
+    self.data = data
+
+    if self.data:
+      self._validate_and_adjust_duration()
 
   def expand(self, pbegin):
     result = (
         pbegin
         | 'ImpulseElement' >> beam.Create(
             [(self.start_ts, self.stop_ts, self.interval)])
-        | 'GenSequence' >> beam.ParDo(ImpulseSeqGenDoFn())
-        | 'MapToTimestamped' >> beam.Map(lambda tt: TimestampedValue(tt, tt)))
+        | 'GenSequence' >> beam.ParDo(ImpulseSeqGenDoFn(self.data)))
+
+    if not self.data:
+      # This step is only to ensure the current PTransform expansion is
+      # compatible with the previous Beam versions.
+      result = (
+          result
+          | 'MapToTimestamped' >> beam.Map(lambda tt: TimestampedValue(tt, tt)))
```

Comment:
Can this one now actually just be the identity function? If that breaks something it is fine to leave, but at this point I think all values should be timestamped, right?

Reply:
Yeah, if the coder changes that causes compat issues. Lets leave it then, thanks